lib/charms/loki_k8s/v0/loki_push_api.py

```diff
@@ -585,79 +586,194 @@ def _validate_relation_by_interface_and_direction(
-            charm_name=data["charm_name"],
+            unit=data.get("unit", ""),
+            charm_name=data.get("charm_name", ""),
         )
 
     @property
     def identifier(self) -> str:
         """Format the topology information into a terse string."""
-        return "{}_{}_{}".format(self.model, self.model_uuid, self.application)
-
-    @property
-    def short_model_uuid(self) -> str:
-        """Obtain the short form of the model uuid."""
-        return self.model_uuid[:7]
+        # This is odd, but may have `None` as a model key
+        return "_".join([str(val) for val in self.as_dict().values()])
 
     @property
-    def scrape_identifier(self) -> str:
-        """Format the topology information into a scrape identifier."""
-        return "juju_{}_{}_{}".format(
-            self.model,
-            self.short_model_uuid,
-            self.application,
+    def promql_labels(self) -> str:
+        """Format the topology information into a verbose string."""
+        return ", ".join(
+            [
+                'juju_{}="{}"'.format(key, value)
+                for key, value in self.as_dict(rename_keys={"charm_name": "charm"}).items()
+            ]
         )
 
-    @property
-    def logql_labels(self) -> str:
-        """Format the topology information into a verbose string."""
-        return 'juju_model="{}", juju_model_uuid="{}", juju_application="{}"'.format(
-            self.model, self.model_uuid, self.application
+    def as_dict(self, rename_keys: Optional[Dict[str, str]] = None) -> OrderedDict:
+        """Format the topology information into a dict.
+
+        Use an OrderedDict so we can rely on the insertion order on Python 3.5 (and 3.6,
+        which still does not guarantee it).
+
+        Args:
+            rename_keys: A dictionary mapping old key names to new key names, which will
+                be substituted when invoked.
+        """
+        ret = OrderedDict(
+            [
+                ("model", self.model),
+                ("model_uuid", self.model_uuid),
+                ("application", self.application),
+                ("unit", self.unit),
+                ("charm_name", self.charm_name),
+            ]
         )
 
-    def as_dict(self, short_uuid: bool = False) -> dict:
-        """Format the topology information into a dict."""
-        return {
-            "model": self.model,
-            "model_uuid": self.model_uuid,
-            "application": self.application,
-            "charm_name": self.charm_name,
-        }
+        ret["unit"] or ret.pop("unit")
+        ret["charm_name"] or ret.pop("charm_name")
+
+        # If a key exists in `rename_keys`, replace the value
+        if rename_keys:
+            ret = OrderedDict(
+                (rename_keys.get(k), v) if rename_keys.get(k) else (k, v) for k, v in ret.items()  # type: ignore
+            )
+
+        return ret
 
-    def as_dict_with_logql_labels(self) -> dict:
+    def as_promql_label_dict(self):
```

Comment:
promql ain't logql?

```suggestion
    def as_logql_label_dict(self):
```

Reply:
It's not, but the content is exactly the same.

We should open an issue to just rename this to `as_query_language_label_dict` or similar.

I left it as `as_promql_label_dict` because the difference between `metadata` in the Loki data bags for alerts is difference enough, and "keep all of this stuff in sync across 3 different libraries" (prometheus_remote_write, prometheus_scrape, loki_push_api) is bad enough without needing to remember the caveats for all of them.